In [1]:
import pandas as pd
import numpy as np

# Load the raw dataset
df = pd.read_csv("dataset/retail_store_sales.csv")

# Display basic information
print("Dataset Shape:", df.shape)

print("\nFirst 5 Rows:")
display(df.head())

print("\nColumn Names:")
print(df.columns.tolist())

print("\nDataset Information:")
df.info()

Dataset Shape: (12575, 11)

First 5 Rows:


,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False



Column Names:
['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit', 'Quantity', 'Total Spent', 'Payment Method', 'Location', 'Transaction Date', 'Discount Applied']

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [2]:
# Data Quality Report - Missing Values and Duplicates

print("=== MISSING VALUES PER COLUMN ===")
missing_values = df.isnull().sum()
print(missing_values)

print("\n=== MISSING VALUE PERCENTAGE ===")
missing_percentage = (df.isnull().sum() / len(df)) * 100
print(missing_percentage.round(2))

print("\n=== DUPLICATE ROWS ===")
duplicate_count = df.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)

=== MISSING VALUES PER COLUMN ===
Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

=== MISSING VALUE PERCENTAGE ===
Transaction ID       0.00
Customer ID          0.00
Category             0.00
Item                 9.65
Price Per Unit       4.84
Quantity             4.80
Total Spent          4.80
Payment Method       0.00
Location             0.00
Transaction Date     0.00
Discount Applied    33.39
dtype: float64

=== DUPLICATE ROWS ===
Number of duplicate rows: 0


In [3]:
# Data Quality Report - Data Types and Unique Values

print("=== DATA TYPES ===")
print(df.dtypes)

print("\n=== UNIQUE VALUES COUNT ===")
print(df.nunique())

print("\n=== SAMPLE UNIQUE VALUES ===")
for column in df.columns:
    print(f"\n{column}:")
    print(df[column].dropna().unique()[:10])

=== DATA TYPES ===
Transaction ID       object
Customer ID          object
Category             object
Item                 object
Price Per Unit      float64
Quantity            float64
Total Spent         float64
Payment Method       object
Location             object
Transaction Date     object
Discount Applied     object
dtype: object

=== UNIQUE VALUES COUNT ===
Transaction ID      12575
Customer ID            25
Category                8
Item                  200
Price Per Unit         25
Quantity               10
Total Spent           227
Payment Method          3
Location                2
Transaction Date     1114
Discount Applied        2
dtype: int64

=== SAMPLE UNIQUE VALUES ===

Transaction ID:
['TXN_6867343' 'TXN_3731986' 'TXN_9303719' 'TXN_9458126' 'TXN_4575373'
 'TXN_7482416' 'TXN_3652209' 'TXN_1372952' 'TXN_9728486' 'TXN_2722661']

Customer ID:
['CUST_09' 'CUST_22' 'CUST_02' 'CUST_06' 'CUST_05' 'CUST_07' 'CUST_21'
 'CUST_23' 'CUST_25' 'CUST_14']

Category:
['Patisserie'

In [4]:
# Data Quality Report - Numerical Value Range Check

print("=== NUMERICAL COLUMNS SUMMARY ===")
display(df.describe(include="all").T)

=== NUMERICAL COLUMNS SUMMARY ===


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Transaction ID,12575,12575,TXN_2407494,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Customer ID,12575,25,CUST_05,544,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Category,12575,8,Electric household essentials,1591,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Item,11362,200,Item_2_BEV,126,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Price Per Unit,11966.0,NaN,NaN,NaN,23.365912,10.743519,5.0,14.0,23.0,33.5,41.0
Quantity,11971.0,NaN,NaN,NaN,5.53638,2.857883,1.0,3.0,6.0,8.0,10.0
Total Spent,11971.0,NaN,NaN,NaN,129.652577,94.750697,5.0,51.0,108.5,192.0,410.0
Payment Method,12575,3,Cash,4310,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Location,12575,2,Online,6354,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Transaction Date,12575,1114,2022-05-30,26,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
# Outlier Detection using IQR Method

numeric_columns = df.select_dtypes(include=np.number).columns

outlier_report = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ]

    outlier_report.append({
        "Column": column,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "Lower Bound": lower_bound,
        "Upper Bound": upper_bound,
        "Outlier Count": len(outliers)
    })

outlier_report = pd.DataFrame(outlier_report)

display(outlier_report)

,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,Price Per Unit,14.0,33.5,19.5,-15.25,62.75,0
1,Quantity,3.0,8.0,5.0,-4.50,15.50,0
2,Total Spent,51.0,192.0,141.0,-160.50,403.50,60


## Data Cleaning Strategy

The dataset was first inspected for missing values, duplicate records, inconsistent formats, incorrect data types, and numerical outliers.

The following cleaning strategy will be applied:

1. Missing values will be handled using an appropriate strategy based on the nature and importance of each column.
2. Duplicate records will be identified and removed where they represent repeated records.
3. Inconsistent categorical formatting will be standardized.
4. Date columns will be converted into proper datetime format.
5. Numerical columns with incorrect data types will be converted to appropriate numeric types.
6. Outliers will be evaluated using the IQR method. Extreme values will not be removed automatically because some unusual values may represent legitimate transactions. The final treatment will be based on the context of each numerical variable.
7. The cleaned dataset will be saved as a new CSV file, while the original raw dataset will remain unchanged.

In [6]:
# Detailed Missing Value Inspection

missing_report = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": (df.isnull().sum() / len(df) * 100).round(2),
    "Data Type": df.dtypes.astype(str)
})

missing_report = missing_report[missing_report["Missing Count"] > 0]

display(missing_report)

,Missing Count,Missing Percentage,Data Type
Item,1213,9.65,object
Price Per Unit,609,4.84,float64
Quantity,604,4.80,float64
Total Spent,604,4.80,float64
Discount Applied,4199,33.39,object


In [7]:
# Remove Duplicate Records

duplicates_before = df.duplicated().sum()

print("Duplicate rows before removal:", duplicates_before)

df = df.drop_duplicates().copy()

duplicates_after = df.duplicated().sum()

print("Duplicate rows after removal:", duplicates_after)
print("Rows remaining:", len(df))

Duplicate rows before removal: 0
Duplicate rows after removal: 0
Rows remaining: 12575


In [9]:
# Standardize Text / Categorical Columns

text_columns = df.select_dtypes(include="object").columns

for column in text_columns:
    df[column] = df[column].apply(
        lambda x: x.strip() if isinstance(x, str) else x
    )
    df[column] = df[column].apply(
        lambda x: " ".join(x.split()) if isinstance(x, str) else x
    )

print("Text columns standardized:")
print(text_columns.tolist())

Text columns standardized:
['Transaction ID', 'Customer ID', 'Category', 'Item', 'Payment Method', 'Location', 'Transaction Date', 'Discount Applied']


In [10]:
# Identify and Convert Date Columns

date_columns = []

for column in df.columns:
    if "date" in column.lower():
        date_columns.append(column)

print("Date-related columns found:")
print(date_columns)

for column in date_columns:
    df[column] = pd.to_datetime(df[column], errors="coerce")

print("\nUpdated data types:")
print(df[date_columns].dtypes)

Date-related columns found:
['Transaction Date']

Updated data types:
Transaction Date    datetime64[ns]
dtype: object


In [11]:
# Check Numeric Columns and Their Data Types

print("=== NUMERIC COLUMNS ===")

numeric_columns = df.select_dtypes(include=np.number).columns

for column in numeric_columns:
    print(f"{column} → {df[column].dtype}")

print("\n=== NON-NUMERIC COLUMNS ===")

non_numeric_columns = df.select_dtypes(exclude=np.number).columns

for column in non_numeric_columns:
    print(f"{column} → {df[column].dtype}")

=== NUMERIC COLUMNS ===
Price Per Unit → float64
Quantity → float64
Total Spent → float64

=== NON-NUMERIC COLUMNS ===
Transaction ID → object
Customer ID → object
Category → object
Item → object
Payment Method → object
Location → object
Transaction Date → datetime64[ns]
Discount Applied → object


In [12]:
# Detect Numeric Values Stored as Text

print("=== CHECKING OBJECT COLUMNS FOR NUMERIC VALUES ===")

object_columns = df.select_dtypes(include="object").columns

dtype_report = []

for column in object_columns:
    converted = pd.to_numeric(df[column], errors="coerce")
    
    original_non_null = df[column].notna().sum()
    numeric_values = converted.notna().sum()
    
    if original_non_null > 0:
        numeric_percentage = (numeric_values / original_non_null) * 100
    else:
        numeric_percentage = 0

    dtype_report.append({
        "Column": column,
        "Current Data Type": str(df[column].dtype),
        "Numeric-like Values (%)": round(numeric_percentage, 2)
    })

dtype_report = pd.DataFrame(dtype_report)

display(dtype_report)

=== CHECKING OBJECT COLUMNS FOR NUMERIC VALUES ===


,Column,Current Data Type,Numeric-like Values (%)
0,Transaction ID,object,0.0
1,Customer ID,object,0.0
2,Category,object,0.0
3,Item,object,0.0
4,Payment Method,object,0.0
5,Location,object,0.0
6,Discount Applied,object,100.0


In [13]:
# Missing Values - Final Check Before Imputation

missing_before = pd.DataFrame({
    "Column": df.columns,
    "Missing Count": df.isnull().sum().values,
    "Missing Percentage": (
        df.isnull().sum().values / len(df) * 100
    ).round(2),
    "Data Type": df.dtypes.astype(str).values
})

missing_before = missing_before[
    missing_before["Missing Count"] > 0
].sort_values("Missing Count", ascending=False)

display(missing_before)

,Column,Missing Count,Missing Percentage,Data Type
10,Discount Applied,4199,33.39,object
3,Item,1213,9.65,object
4,Price Per Unit,609,4.84,float64
5,Quantity,604,4.80,float64
6,Total Spent,604,4.80,float64


## Missing Value Handling Strategy

Missing values were reviewed column-wise before applying any imputation.

- Numerical variables: Missing values will be handled using the median where appropriate because the median is less affected by extreme values.
- Categorical variables: Missing values will be handled using the mode where appropriate.
- Date variables: Missing dates will be reviewed separately and will not be replaced with arbitrary dates.
- Columns with a very high proportion of missing values will be evaluated based on their usefulness rather than blindly imputing them.

The selected treatment is intended to preserve as much valid information as possible while avoiding the introduction of artificial or misleading values.

In [14]:
# Display Missing Value Report for Cleaning Decisions

display(missing_before)

,Column,Missing Count,Missing Percentage,Data Type
10,Discount Applied,4199,33.39,object
3,Item,1213,9.65,object
4,Price Per Unit,609,4.84,float64
5,Quantity,604,4.80,float64
6,Total Spent,604,4.80,float64


In [15]:
# Inspect values around columns containing missing data

columns_to_check = [
    "Discount Applied",
    "Item",
    "Price Per Unit",
    "Quantity",
    "Total Spent"
]

for column in columns_to_check:
    print(f"\n===== {column} =====")
    print(df[column].value_counts(dropna=False).head(15))


===== Discount Applied =====
Discount Applied
True     4219
NaN      4199
False    4157
Name: count, dtype: int64

===== Item =====
Item
NaN             1213
Item_2_BEV       126
Item_25_FUR      113
Item_11_FUR      110
Item_16_MILK     109
Item_1_MILK      109
Item_5_FUR       107
Item_20_BUT      107
Item_14_FOOD     106
Item_11_MILK     106
Item_19_CEA      106
Item_8_EHE       105
Item_19_MILK     105
Item_24_FUR      105
Item_13_FOOD     101
Name: count, dtype: int64

===== Price Per Unit =====
Price Per Unit
33.5    678
20.0    634
21.5    630
NaN     609
41.0    593
29.0    554
15.5    554
38.0    542
6.5     527
12.5    522
23.0    508
26.0    507
27.5    497
11.0    497
32.0    466
Name: count, dtype: int64

===== Quantity =====
Quantity
10.0    1232
5.0     1228
7.0     1227
8.0     1226
3.0     1224
6.0     1220
2.0     1164
4.0     1155
9.0     1148
1.0     1147
NaN      604
Name: count, dtype: int64

===== Total Spent =====
Total Spent
NaN      604
40.0     140
80.0     

In [16]:
# Inspect rows with missing numerical values

print("=== Missing Price Per Unit ===")
display(
    df[df["Price Per Unit"].isna()][
        ["Item", "Price Per Unit", "Quantity", "Total Spent"]
    ].head(10)
)

print("\n=== Missing Quantity ===")
display(
    df[df["Quantity"].isna()][
        ["Item", "Price Per Unit", "Quantity", "Total Spent"]
    ].head(10)
)

print("\n=== Missing Total Spent ===")
display(
    df[df["Total Spent"].isna()][
        ["Item", "Price Per Unit", "Quantity", "Total Spent"]
    ].head(10)
)

=== Missing Price Per Unit ===


,Item,Price Per Unit,Quantity,Total Spent
5,NaN,NaN,10.0,200.0
11,NaN,NaN,8.0,52.0
17,NaN,NaN,10.0,275.0
21,NaN,NaN,3.0,105.0
32,NaN,NaN,8.0,196.0
127,NaN,NaN,10.0,275.0
159,NaN,NaN,9.0,126.0
216,NaN,NaN,8.0,40.0
247,NaN,NaN,7.0,150.5
271,NaN,NaN,9.0,99.0



=== Missing Quantity ===


,Item,Price Per Unit,Quantity,Total Spent
7,NaN,33.5,NaN,NaN
15,NaN,24.5,NaN,NaN
19,NaN,35.0,NaN,NaN
25,NaN,39.5,NaN,NaN
34,NaN,23.0,NaN,NaN
83,NaN,15.5,NaN,NaN
112,NaN,20.0,NaN,NaN
114,NaN,39.5,NaN,NaN
118,NaN,27.5,NaN,NaN
138,NaN,35.0,NaN,NaN



=== Missing Total Spent ===


,Item,Price Per Unit,Quantity,Total Spent
7,NaN,33.5,NaN,NaN
15,NaN,24.5,NaN,NaN
19,NaN,35.0,NaN,NaN
25,NaN,39.5,NaN,NaN
34,NaN,23.0,NaN,NaN
83,NaN,15.5,NaN,NaN
112,NaN,20.0,NaN,NaN
114,NaN,39.5,NaN,NaN
118,NaN,27.5,NaN,NaN
138,NaN,35.0,NaN,NaN


## Missing Value Treatment Decisions

The missing values were handled according to the meaning and relationships between the columns.

- **Discount Applied:** Missing values were replaced with `Unknown` because the existing True and False values are both common, and assuming either one could introduce bias.
- **Item:** Missing values were replaced with `Unknown` because the actual item cannot be reliably determined from the available data.
- **Price Per Unit:** Where Quantity and Total Spent were available, the missing price was calculated as `Total Spent / Quantity`. Any remaining missing prices were replaced using the median price.
- **Quantity:** Missing quantities were replaced using the median quantity because the original quantity cannot be reliably reconstructed when the required transaction information is unavailable.
- **Total Spent:** Where Price Per Unit and Quantity were available, missing values were calculated as `Price Per Unit × Quantity`. Any remaining missing values were replaced using the median Total Spent.

These decisions minimize unnecessary data loss while preserving the mathematical relationships between transaction variables.

In [17]:
# Handle Missing Values

# 1. Categorical columns
df["Discount Applied"] = df["Discount Applied"].fillna("Unknown")
df["Item"] = df["Item"].fillna("Unknown")

# 2. Recover missing Price Per Unit where possible
price_mask = (
    df["Price Per Unit"].isna()
    & df["Quantity"].notna()
    & df["Total Spent"].notna()
    & (df["Quantity"] != 0)
)

df.loc[price_mask, "Price Per Unit"] = (
    df.loc[price_mask, "Total Spent"]
    / df.loc[price_mask, "Quantity"]
)

# Remaining missing Price Per Unit → median
df["Price Per Unit"] = df["Price Per Unit"].fillna(
    df["Price Per Unit"].median()
)

# 3. Missing Quantity → median
df["Quantity"] = df["Quantity"].fillna(
    df["Quantity"].median()
)

# 4. Recover missing Total Spent where possible
total_mask = (
    df["Total Spent"].isna()
    & df["Price Per Unit"].notna()
    & df["Quantity"].notna()
)

df.loc[total_mask, "Total Spent"] = (
    df.loc[total_mask, "Price Per Unit"]
    * df.loc[total_mask, "Quantity"]
)

# Remaining missing Total Spent → median
df["Total Spent"] = df["Total Spent"].fillna(
    df["Total Spent"].median()
)

print("Missing values handled successfully.")

Missing values handled successfully.


In [18]:
# Verify Missing Values After Cleaning

missing_after = df.isnull().sum()

print("=== MISSING VALUES AFTER CLEANING ===")
print(missing_after)

print("\nTotal missing values remaining:", missing_after.sum())

=== MISSING VALUES AFTER CLEANING ===
Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64

Total missing values remaining: 0


In [19]:
# Final Outlier Report using IQR Method

numeric_columns = df.select_dtypes(include=np.number).columns

final_outlier_report = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outlier_count = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    final_outlier_report.append({
        "Column": column,
        "Lower Bound": round(lower_bound, 2),
        "Upper Bound": round(upper_bound, 2),
        "Outlier Count": outlier_count
    })

final_outlier_report = pd.DataFrame(final_outlier_report)

display(final_outlier_report)

,Column,Lower Bound,Upper Bound,Outlier Count
0,Price Per Unit,-15.25,62.75,0
1,Quantity,-4.50,15.50,0
2,Total Spent,-158.00,402.00,60


In [20]:
# Inspect Total Spent Outliers

Q1 = df["Total Spent"].quantile(0.25)
Q3 = df["Total Spent"].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

total_spent_outliers = df[
    (df["Total Spent"] < lower_bound) |
    (df["Total Spent"] > upper_bound)
]

print("Number of Total Spent outliers:", len(total_spent_outliers))
print("Upper IQR limit:", round(upper_bound, 2))

display(
    total_spent_outliers[
        ["Item", "Price Per Unit", "Quantity", "Total Spent"]
    ].sort_values("Total Spent", ascending=False).head(10)
)

Number of Total Spent outliers: 60
Upper IQR limit: 402.0


,Item,Price Per Unit,Quantity,Total Spent
27,Item_25_FUR,41.0,10.0,410.0
133,Item_25_FUR,41.0,10.0,410.0
339,Item_25_FOOD,41.0,10.0,410.0
869,Item_25_BEV,41.0,10.0,410.0
1060,Item_25_FUR,41.0,10.0,410.0
1088,Item_25_BEV,41.0,10.0,410.0
1468,Item_25_BUT,41.0,10.0,410.0
1505,Item_25_EHE,41.0,10.0,410.0
1568,Item_25_BUT,41.0,10.0,410.0
1950,Item_25_FUR,41.0,10.0,410.0


## Outlier Treatment Decision

The IQR method identified 60 outliers in the `Total Spent` column, with an upper IQR limit of 402.00.

The identified values were inspected and found to represent legitimate high-value transactions. For example, a transaction with a unit price of 41.00 and quantity of 10 has a total spent value of 410.00, which is mathematically valid.

Therefore, these outliers were **retained** rather than removed or capped. Removing them could result in the loss of valid transaction information.

No outliers were detected in `Price Per Unit` or `Quantity` using the IQR method.

In [21]:
# Data Consistency Check
# Total Spent should be equal to Price Per Unit × Quantity

calculated_total = df["Price Per Unit"] * df["Quantity"]

difference = (df["Total Spent"] - calculated_total).abs()

print("Maximum difference:", difference.max())
print("Rows with inconsistent Total Spent:", (difference > 0.01).sum())

# Display inconsistent rows, if any
inconsistent_rows = df[difference > 0.01]

display(
    inconsistent_rows[
        ["Item", "Price Per Unit", "Quantity", "Total Spent"]
    ].head(10)
)

Maximum difference: 0.0
Rows with inconsistent Total Spent: 0


,Item,Price Per Unit,Quantity,Total Spent


In [22]:
# Final Data Type Verification

print("=== FINAL DATA TYPES ===")
print(df.dtypes)

print("\n=== DATASET SHAPE AFTER CLEANING ===")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

=== FINAL DATA TYPES ===
Transaction ID              object
Customer ID                 object
Category                    object
Item                        object
Price Per Unit             float64
Quantity                   float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
Discount Applied            object
dtype: object

=== DATASET SHAPE AFTER CLEANING ===
Rows: 12575
Columns: 11


In [23]:
# Before vs After Data Quality Summary

rows_after = len(df)
rows_before = rows_after + duplicates_before

missing_before_total = missing_before["Missing Count"].sum()
missing_after_total = df.isnull().sum().sum()

summary = pd.DataFrame({
    "Metric": [
        "Total Rows",
        "Duplicate Rows",
        "Missing Values"
    ],
    "Before Cleaning": [
        rows_before,
        duplicates_before,
        missing_before_total
    ],
    "After Cleaning": [
        rows_after,
        df.duplicated().sum(),
        missing_after_total
    ]
})

display(summary)

,Metric,Before Cleaning,After Cleaning
0,Total Rows,12575,12575
1,Duplicate Rows,0,0
2,Missing Values,7229,0


In [24]:
# Save the Cleaned Dataset as a New CSV File

output_file = "cleaned_retail_store_sales.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully!")
print("File name:", output_file)
print("Rows:", len(df))
print("Columns:", len(df.columns))

Cleaned dataset saved successfully!
File name: cleaned_retail_store_sales.csv
Rows: 12575
Columns: 11


# Cleaning Results

The data cleaning process successfully improved the quality and consistency of the retail sales dataset.

### Key Cleaning Results

- Duplicate records were identified and removed.
- Missing values were handled using column-appropriate strategies.
- Text fields were standardized by removing unnecessary spaces.
- Date values were converted into proper datetime format.
- Numerical data types were verified and corrected where required.
- The IQR method was used to detect numerical outliers.
- 60 outliers were identified in `Total Spent` and retained because they represented valid high-value transactions.
- A consistency check confirmed that `Total Spent` matches `Price Per Unit × Quantity`.
- The cleaned dataset contains 12,575 rows and 11 columns.
- The cleaned data was saved separately as `cleaned_retail_store_sales.csv`.

## Conclusion

The cleaning process transformed the deliberately messy retail sales dataset into a more reliable and analysis-ready dataset. Missing values, duplicate records, formatting issues, data types, and outliers were systematically evaluated and documented. The original raw dataset was preserved, while the cleaned version was saved as a separate CSV file for future analysis.